### ISSUE TITLE
Add tweener that waits for a signal to be emitted

### ISSUE THREAD

**Issue by Marigem** [godotengine/godot-proposals#7337]
### Describe the project you are working on

2D platforming RPG with dynamic character cutscenes.

### Describe the problem or limitation you are having in your project

The current signal system without callbacks (Godot 3.x) makes it difficult to create a large sequence of events, which include lots of timers and delays.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a "wait_for_signal" method to the SceneTreeTweens that creates a Tweener which finishes when the signal is emitted, or when a certain amount of time passes (a timeout parameter). Using custom_step on the tween should finish the Tweener immediately.
When the target signal is emitted, the Tweener's finish signal should be emitted on the physics or idle frame, depending on the tween's process_mode, if that's consistent with already existing tweeners.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The result would look something like:
```
func cutscene() -> void:
    cs_tween = create_tween()
    cs_tween.set_process_mode(Tween.TWEEN_PROCESS_PHYSICS)
    cs_tween.tween_callback(character, "go_to_pos", [some_vec2])
    cs_tween.wait_for_signal(character, "go_to_pos_finished", 10.0)
    cs_tween.tween_interval(0.25)
    cs_tween.tween_callback(character, "turn_around")
    cs_tween.tween_interval(0.25)
    cs_tween.tween_callback(character.anim_player, "play", ["some_animation"])
    cs_tween.wait_for_signal(character.anim_player, "animation_finished", 10.0)
    ...
```
Very readable, editable and can be easily cancelled at any time.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

There current options are:

**1 - Safe option: Create a function every time you need to wait for a signal, and connect the signal to it.**
- This can add up when there are many little delays in between every action. For example: Character goes to position, waits 0.25sec, turns around, does something else, some variable is changed, function is called etc. The code will look something like:
```
func cutscene_pt1() -> void:
    character.go_to_pos(some_vec2)
    character.connect("go_to_pos_finished", self, "cutscene_pt2")
    
func cutscene_pt2() -> void:
    get_tree().create_timer(0.25).connect("timeout", self, "cutscene_pt3")
    
func cutscene_pt3() -> void:
    character.turn_around()
    get_tree().create_timer(0.25).connect("timeout", self, "cutscene_pt4")
    
func cutscene_pt4() -> void:
    character.anim_player.play("some_animation")
    character.anim_player.connect("animation_finished", self, "cutscene_pt5")
    
func cutscene_pt5(_anim: String) -> void:
    ...
```
The problem is that making any edits, especially to the order of events will be very difficult.

**2 - Unsafe option: Use yield to wait for every signal. this forces us to make sure that the player can't pause and reload or do anything that will make the yielded function throw an Error:**

```
func cutscene():
    character.go_to_pos(some_vec2)
    yield(character, "go_to_pos_finished")
    yield(get_tree().create_timer(0.25), "timeout")
    character.turn_around()
    yield(get_tree().create_timer(0.25), "timeout")
    character.anim_player.play("some_animation")
    yield(character.anim_player, "animation_finished")
    ...
```
While this is more readable and editable, the problem here is that it's not cancellable. It's an unskippable sequence of events

**3 - Complex: Create a custom class that can handle every possible kind of event.**

### Is there a reason why this should be core and not an add-on in the asset library?

SceneTreeTween is not extensible to my knowledge.

**Comment by dalexeev** [comment]
In my opinion, this makes sense, since using GDScript's `await` with `Tween` is less convenient (and besides makes the function a coroutine):

```gdscript
tween.tween_property(...)
await tween.finished
tween.stop()
await some_signal
tween.tween_property(...)
```

But perhaps we should think about a more consistent name (should start with `tween_`). I suggest `tween_signal()` or `tween_await_signal()`. Also, it should be a single argument of type `Signal`, not `Object` + a string. This is consistent with the `tween_callback()` and `tween_method()` methods, which accept a `Callable` rather than an `Object` + a string.

**Comment by Marigem** [comment]
Yes, the name was just a placeholder and it makes more sense for it to start with `tween_`.
As for the arguments, in godot 3.x `tween_callback` and `tween_method` accept an `Object` + `String`. While in godot 4 they accept a `Callable`. So that would be the same in this case as well.

**Comment by KoBeWi** [comment]
I was already considering adding something like this, but I use Tweens only for simple animations. I have a custom system for cutscenes, which already has this functionality. But I agree that it's something useful.

My proposed name is `tween_await()`. The timeout wouldn't be an optional argument, but a method `set_timeout()`, to be consistent with other Tweeners.

**Comment by KoBeWi** [comment]
No. At best to 3.6, but someone needs to do the backport.

**Comment by chesedcore** [comment]
folks you have no idea how awesome this change is, as i see the documentation i am literally crying at how useful this is, i'm just here to thank you all from the bottom of my heart for this amazing new feature
